File: Week3/insertsort.py
import sys


def insertion_sort(arr):
    comparisons = 0
    shifts = 0
    n = len(arr)

    for i in range(1, n):
        key = arr[i]
        j = i - 1

        while j >= 0:
            comparisons += 1
            if arr[j] > key:
                arr[j + 1] = arr[j]
                shifts += 1
                j -= 1
            else:
                break

        arr[j + 1] = key

    return comparisons, shifts


def main():
    input_data = sys.stdin.read().strip().split()
    t = int(input_data[0])
    index = 1

    outputs = []

    for _ in range(t):
        n = int(input_data[index])
        index += 1

        arr = list(map(int, input_data[index:index + n]))
        index += n

        comparisons, shifts = insertion_sort(arr)

        outputs.append(" ".join(map(str, arr)))
        outputs.append(f"comparisons = {comparisons}")
        outputs.append(f"shifts = {shifts}")

    print("\n".join(outputs))


if __name__ == "__main__":
    main()

File: Week3/insertsort.java
import java.util.*;
class insertsort{
	static class Result{
	int comparisons;
	int shifts;
	}

	static Result insertionSort(int[] arr){
		Result res = new Result();
		int n = arr.length;

		for(int i = 1, i < n; i++){
			int key = arr[i];
			int j = i-1;

			while(j >= 0){
				res.comparisons++;
				if(arr[j] > key){
					arr[j + 1] = arr[j];
					res.shifts++;
					j--;
				} else {
					break;
				}
			}
			arr[j + 1] = key;
		}
		return res;
	}

	public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();

        while (t > 0) {
            int n = sc.nextInt();
            int[] arr = new int[n];
            System.out.println("");
            for (int i = 0; i < n; i++) {
                arr[i] = sc.nextInt();
            }

            Result res = insertionSort(arr);

            for (int x : arr) {
                System.out.print(x + " ");
            }
            System.out.println();
            System.out.println("comparisons = " + res.comparisons);
            System.out.println("shifts = " + res.shifts);
            t--;
        }
        sc.close();
    }
}